### PR TITLE
[BEAM-2174] Update CoderRegistry to allow creating coders through CoderFactory for a wider range of types

### PR DESCRIPTION
**Pull request by lukecwik**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
This removes getInstanceComponents and hard codes the inference for commonly "known" types into Create removing the additionally complexity that was brought along with it.

Also:
* @DefaultCoder to become a CoderFactory
* Created a ProtoCoderFactory which delegates to the ProtoCoder for all Message types
* Created a SerializableCoderFactory which delegates to the SerializableCoder for all Serializable types
* Created a WritableCoderFactory which delegates to the Hadoop WritableCoder for all Hadoop Writable types
* Renamed getDefaultCoder to getCoder on CoderRegistry

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @davorbonaci @kennknowles

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java [hunk 1]
```diff
@@ -284,7 +284,7 @@ public PCollectionTuple expand(
       return input
           .getPipeline()
           .getCoderRegistry()
-          .getDefaultCoder(output.getTypeDescriptor(), fn.getInputTypeDescriptor(), inputCoder);
+          .getCoder(output.getTypeDescriptor(), fn.getInputTypeDescriptor(), inputCoder);
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: Separate the rename of `getDefaultCoder` -> `getCoder` into a commit, so it can be ignored for real review.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java [hunk 2]
```diff
@@ -139,18 +140,39 @@
...
     @Override
-    public <T> Coder<T> getCoder(TypeDescriptor<T> typeDescriptor) {
-      // This is a downcast from `? super T` to T. However, because
-      // it comes from a TypeDescriptor<T>, the class object itself
-      // is the same so the supertype in question shares the same
-      // generated AvroCoder schema.
-      @SuppressWarnings("unchecked")
-      Class<T> rawType = (Class<T>) typeDescriptor.getRawType();
-      return AvroCoder.of(rawType);
+    public <T> Coder<T> create(TypeDescriptor<T> typeDescriptor,
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java]: Prefer `getCoder(...)` or `coderFor(...)` instead of `create(...)` because whether or not it actually creates a new coder is an implementation detail.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/ByteArrayCoder.java [hunk 3]
```diff
@@ -47,13 +46,6 @@ public static ByteArrayCoder of() {
     return INSTANCE;
   }
 
-  /**
-   * Returns an empty list. {@link ByteArrayCoder} has no components.
-   */
-  public static <T> List<Object> getInstanceComponents(T ignored) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/ByteArrayCoder.java]: Separate deletion of `getInstanceComponents` to its own commit, too.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java [hunk 4]
```diff
@@ -34,18 +32,14 @@
   private CoderFactories() { } // Static utility class
 
   /**
-   * Creates a {@link CoderFactory} built from particular static methods of a class that
-   * implements {@link Coder}.
-   *
-   * <p>The class must have the following static methods:
+   * Creates a {@link CoderFactory} for a given raw type which is able to build a coder from a
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java]: The term "raw type" refers to the relationship between an actual type and a class. The phrasing is also ambiguous - the `CoderFactory` should work for more than raw types, that's the whole point.

Since it is now just one static method, how about just shortening it to:

`"Creates a {@link CoderFactory} from a class's {@code static <T> Coder<T> of(TypeDescriptor<T>, List<Coder<?>>}) method."`

You probably want to switch from varargs to list here. That will then allow a single factory to provide coders for types with various numbers of type args.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java [hunk 5]
```diff
@@ -69,30 +59,33 @@ private CoderFactories() { } // Static utility class
...
+  public static <T> CoderFactory forCoder(TypeDescriptor<T> type, Coder<T> coder) {
+    return new CoderFactoryForCoder<>(type, coder);
   }
 
   /**
    * See {@link #fromStaticMethods} for a detailed description
    * of the characteristics of this {@link CoderFactory}.
    */
-  private static class CoderFactoryFromStaticMethods implements CoderFactory {
+  private static class CoderFactoryFromStaticMethods<T> implements CoderFactory {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java]: This shouldn't be parameterized. The `<T>` comes as a generic param to the method below.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java [hunk 6]
```diff
@@ -197,92 +169,43 @@ private Method getFactoryMethod(Class<?> coderClazz) {
...
-   * {@link CoderFactory}.
+   * See {@link #forCoder} for a detailed description of this {@link CoderFactory}.
    */
   private static class CoderFactoryForCoder<T> implements CoderFactory {
     private final Coder<T> coder;
+    private final TypeDescriptor<T> type;
 
-    public CoderFactoryForCoder(Coder<T> coder) {
+    public CoderFactoryForCoder(TypeDescriptor<T> type, Coder<T> coder){
+      this.type = type;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java]: `coder.getEncodedTypeDescriptor()`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 7]
```diff
@@ -57,191 +55,106 @@
...
- *       {@link #getFallbackCoderProvider()} and {@link #setFallbackCoderProvider}. Multiple
- *       fallbacks can be chained together using {@link CoderProviders#firstOf}.
- * </ol>
+ * <p>Creation of the {@link Coder} is delegated to one of the many registered
+ * {@link CoderFactory coder factories} based upon the registration order.
+ *
+ * <p>By default, the {@link CoderFactory coder factory} registration order is as follows:
+ * <ul>
+ *   <li>Coder factories registered programmatically with
+ *       {@link CoderRegistry#registerCoderFactory(CoderFactory)}.</li>
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Naming fun: shorten to `register` and all the overloads can be the same, though you could have `registerCoderFactory` and `registerCoder` separate, too.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 8]
```diff
@@ -57,191 +55,106 @@
...
+        TimestampedValue.class, TimestampedValue.TimestampedValueCoder.class));
+    codersToRegister.add(CoderFactories.fromStaticMethods(Void.class, VoidCoder.class));
+    codersToRegister.add(CoderFactories.fromStaticMethods(byte[].class, ByteArrayCoder.class));
+    codersToRegister.add(
+        CoderFactories.forCoder(
+            TypeDescriptor.of(IntervalWindow.class), IntervalWindow.getCoder()));
 
     // Enumerate all the CoderRegistrars in a deterministic order, adding all coders to register
-    Set<CoderRegistrar> registrars = Sets.newTreeSet(ObjectsClassComparator.INSTANCE);
+    Set<CoderFactoryRegistrar> registrars = Sets.newTreeSet(ObjectsClassComparator.INSTANCE);
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Using a deterministic order could cause accidental success, leading to mystery breakage later on. Is there another way to regulate duplicate registrations?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 9]
```diff
@@ -57,191 +55,106 @@
...
-   * <p>This method will be called to decompose a value during the {@link Coder} inference process,
-   * to automatically choose {@link Coder Coders} for the components.
-   *
-   * @param clazz the class of objects to be encoded
-   * @param coderClazz a class with static factory methods to provide {@link Coder Coders}
+   * <p>See {@link CoderFactories} for common {@link CoderFactory} patterns.
    */
-  public void registerCoder(Class<?> clazz, Class<?> coderClazz) {
-    registerCoder(clazz, CoderFactories.fromStaticMethods(coderClazz));
+  public void registerCoderFactory(CoderFactory coderFactory) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: This rename could also be its own commit, for clarity.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 10]
```diff
@@ -57,191 +55,106 @@
...
-
-  /**
-   * Register the provided {@link Coder} for encoding all values of the specified {@code Class}.
-   * This will override prior registrations if any exist.
-   *
-   * <p>Not for use with generic rawtypes. Instead, register a {@link CoderFactory} via
-   * {@link #registerCoder(Class, CoderFactory)} or ensure your {@code Coder} class has the
-   * appropriate static methods and register it directly via {@link #registerCoder(Class, Class)}.
-   */
-  public <T> void registerCoder(Class<T> rawClazz, Coder<T> coder) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Don't delete, as mentioned above.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 11]
```diff
@@ -250,30 +163,30 @@ public void registerCoder(Class<?> clazz, CoderFactory coderFactory) {
...
+  public <InputT, OutputT> Coder<OutputT> getCoder(
       TypeDescriptor<OutputT> typeDescriptor,
       TypeDescriptor<InputT> inputTypeDescriptor,
       Coder<InputT> inputCoder)
       throws CannotProvideCoderException {
     checkArgument(typeDescriptor != null);
     checkArgument(inputTypeDescriptor != null);
     checkArgument(inputCoder != null);
-    return getDefaultCoder(
+    return createCoderFromTypeDescriptor(
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Use `getCoderFromTypeDescriptor` not `create`. As before, creation versus caching, etc, is an implementation detail.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 12]
```diff
@@ -250,30 +163,30 @@ public void registerCoder(Class<?> clazz, CoderFactory coderFactory) {
...
         typeDescriptor, getTypeToCoderBindings(inputTypeDescriptor.getType(), inputCoder));
   }
 
   /**
    * Returns the {@link Coder} to use on elements produced by this function, given the {@link Coder}
    * used for its input elements.
    */
-  public <InputT, OutputT> Coder<OutputT> getDefaultOutputCoder(
+  public <InputT, OutputT> Coder<OutputT> getOutputCoder(
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Later, it might be nice to remove this and instead give utility methods for building type-to-coder contexts from various things rather than mysterious methods here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 13]
```diff
@@ -698,84 +485,27 @@ private static boolean isNullOrEmpty(Collection<?> c) {
...
-    LOG.debug("Default coder for {}: {}", typeDescriptor, defaultCoder);
-    @SuppressWarnings("unchecked")
-    Coder<T> result = (Coder<T>) defaultCoder;
-    return result;
-  }
+  private LinkedList<CoderFactory> coderFactories;
 
   /**
-   * Returns the {@link Coder} to use by default for values of the given type,
+   * Creates a {@link Coder} to use by default for values of the given type,
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Returns

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 14]
```diff
@@ -788,74 +518,66 @@ private CoderFactory getDefaultCoderFactory(Class<?> clazz) throws CannotProvide
...
-      Class<?> rawClazz = (Class<?>) type.getRawType();
-      return getDefaultCoderFromAnnotation(rawClazz);
-    } catch (CannotProvideCoderException exc) {
-      annotationException = exc;
+  /**
+   * Attempts to create a {@link Coder} from any registered {@link CoderFactory} returning
+   * the first successfully created instance.
+   */
+  private Coder<?> createCoderFromFactories(
+      TypeDescriptor typeDescriptor, List<Coder<?>> typeArgumentCoders)
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Either `<T> Coder<T> getCoderFromFactories(TypeDescriptor<T>, ...)` or `Coder<?> getCoderFromFactories(TypeDescriptor<?>, ...)` no raw type.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 15]
```diff
@@ -788,74 +518,66 @@ private CoderFactory getDefaultCoderFactory(Class<?> clazz) throws CannotProvide
...
-        .append(". Correct one of the following root causes:");
-
-    messageBuilder
-        .append("\n  Building a Coder using a registered CoderFactory failed: ")
-        .append(factoryException.getMessage());
-
-    messageBuilder
-        .append("\n  Building a Coder from the @DefaultCoder annotation failed: ")
-        .append(annotationException.getMessage());
+        .append("Unable to provide a default Coder for ").append(typeDescriptor)
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: I guess this error message is a bit long now that there are not multiple causes possible. The registered factories should be listed at least, though it is definitely too much to list all of the reasons. That's a major loss of usability, though I don't see a way around it.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Create.java [hunk 16]
```diff
@@ -633,11 +650,48 @@ public T apply(TimestampedValue<T> input) {
...
+        + Create.class.getSimpleName()
+        + ":"
+        + " For their common class, no coder could be provided."
+        + " Based on their values, no coder could be inferred.");
+  }
+
+  /**
+   * Attempt to infer the type for some very common Apache Beam parameterized types.
+   *
+   * <p>TODO: Instead, build a TypeDescriptor so that the {@link CoderRegistry} is invoked
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Create.java]: I think we should address this TODO here

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderFactoriesTest.java [hunk 17]
```diff
@@ -38,48 +42,53 @@
    * exercise CoderFactoryFromStaticMethods.
    */
   @Test
-  public void testAtomicCoderClassFactories() {
-    checkAtomicCoderFactory(StringUtf8Coder.class, StringUtf8Coder.of());
-    checkAtomicCoderFactory(DoubleCoder.class, DoubleCoder.of());
-    checkAtomicCoderFactory(ByteArrayCoder.class, ByteArrayCoder.of());
+  public void testAtomicCoderClassFactories() throws Exception {
+    checkAtomicCoderFactory(String.class, StringUtf8Coder.class, StringUtf8Coder.of());
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderFactoriesTest.java]: What is being checked? This test is not readable. Not your fault, but the method adds nothing; I'd inline it.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java [hunk 18]
```diff
@@ -74,28 +72,11 @@
   private static class NotSerializableClass { }
 
   @Test
-  public void testSerializableFallbackCoderProvider() throws Exception {
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: There should still be an analogous test of falling back to `Serializable` for anything not registered.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java [hunk 19]
```diff
@@ -74,28 +72,11 @@
...
-    CoderRegistry registry = CoderRegistry.createDefault();
-    registry.setFallbackCoderProvider(SerializableCoder.PROVIDER);
-    Coder<?> serializableCoder = registry.getDefaultCoder(SerializableClass.class);
-
-    assertEquals(serializableCoder, SerializableCoder.of(SerializableClass.class));
-  }
-
-  @Test
-  public void testAvroFallbackCoderProvider() throws Exception {
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: Ditto for checking that the user has the ability to change the fallback / insert a fallback at higher priority that `SerializableCoder`.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java [hunk 20]
```diff
@@ -74,28 +72,11 @@
...
-
-    assertEquals(avroCoder, AvroCoder.of(NotSerializableClass.class));
-  }
-
-  @Test
   public void testRegisterInstantiatedCoder() throws Exception {
     CoderRegistry registry = CoderRegistry.createDefault();
-    registry.registerCoder(MyValue.class, MyValueCoder.of());
-    assertEquals(registry.getDefaultCoder(MyValue.class), MyValueCoder.of());
+    registry.registerCoderFactory(
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: Restore prior method, as mentioned above.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java [hunk 21]
```diff
@@ -121,44 +102,35 @@ public void verifyDeterministic() throws NonDeterministicException {}
   }
 
   @Test
-  public void testRegisterInstantiatedCoderInvalidRawtype() throws Exception {
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: Yay! Deleting this test basically summarizes the feature :-)

File: sdks/java/core/src/test/java/org/apache/beam/sdk/values/TypedPValueTest.java [hunk 22]
```diff
@@ -77,14 +78,10 @@ public void testUntypedOutputTupleTagGivesActionableMessage() {
 
     thrown.expect(IllegalStateException.class);
     thrown.expectMessage("No Coder has been manually specified");
-    thrown.expectMessage(
-        containsString("Building a Coder using a registered CoderFactory failed"));
-    thrown.expectMessage(
-        containsString("Building a Coder from the @DefaultCoder annotation failed"));
-    thrown.expectMessage(
-        containsString("Building a Coder from the fallback CoderProvider failed"));
+    thrown.expectMessage("Building a Coder using a registered CoderFactory failed");
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/values/TypedPValueTest.java]: This error message is not even close to actionable. We have to find a way to tell the user what they can do to improve things.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java (hunk 5 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java (hunk 6 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java]: Currently we allow a user to say this coder is good for this type descriptor, do we expect coders are only good for a specific type of descriptor and not something else based upon inheritance like I have an AnimalCoder and I want it to handle my Dog type since it will only encode it as an Animal?

Also, what value does getEncodedTypeDescriptor add on a Coder (since it seems like a good candidate to deprecate/remove and internalize the logic to simplify Coders)?
(Only found one usage in ParDo related to state specs and seems to be related to trying to get the type to coder bindings which we should be able to get again or keep from earlier).

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 8 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: There are no duplicates now, its just a deterministic order so your application doesn't depend on the order in which service files are found on the classpath or the order in which entries appear in those service files. Not have a deterministic order seems like the application will cause accidental success.

Duplicate registrations could be treated as a priority bump so if you register the same coder factory multiple times all it does is move it to the front of the list. WDYT?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 11 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 12 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: I don't think people will use these methods because they are too complicated and are really internal details of how we infer coder propagation for a couple of transforms.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 13 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 14 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 15 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: I was thinking that I could do a follow up PR to address this by actually listing all the failing reasons and collapsing the default into a single coderfactory so that there weren't 50 reasons like its not a bitset, not a byte, not a string failures.

Then the error message would be:
User registered coder factories
Well known coder types
Service loader coder factories (@DefaultCoder, WritableCoder, ProtoCoder)

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java (hunk 19 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: Avro is one that the user shouldn't register since it will erroneously say that it can encode almost anything even if it can't.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java (hunk 18 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: I added this test to SerializableCoder which declares the CoderFactory

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java (hunk 20 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 10 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 7 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: I made the registerYYY methods instead of overloads

File: examples/java/src/test/java/org/apache/beam/examples/complete/TfIdfTest.java [hunk 23]
```diff
@@ -48,7 +50,8 @@
   @Category(ValidatesRunner.class)
   public void testTfIdf() throws Exception {
 
-    pipeline.getCoderRegistry().registerCoder(URI.class, StringDelegateCoder.of(URI.class));
+    pipeline.getCoderRegistry().registerCoderFactory(
+        CoderFactories.forCoder(TypeDescriptor.of(URI.class), StringDelegateCoder.of(URI.class)));
```
Comment (lukecwik) [inline examples/java/src/test/java/org/apache/beam/examples/complete/TfIdfTest.java]: Done

File: examples/java/src/main/java/org/apache/beam/examples/complete/TfIdf.java [hunk 24]
```diff
@@ -409,7 +411,8 @@ public void processElement(ProcessContext c) {
   public static void main(String[] args) throws Exception {
     Options options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
     Pipeline pipeline = Pipeline.create(options);
-    pipeline.getCoderRegistry().registerCoder(URI.class, StringDelegateCoder.of(URI.class));
+    pipeline.getCoderRegistry().registerCoderFactory(
```
Comment (lukecwik) [inline examples/java/src/main/java/org/apache/beam/examples/complete/TfIdf.java]: Done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java (hunk 19 above)
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: It will crash in the middle of generating a schema, right? Or, it might return a nondeterministic coder which will later fail at a GBK. But if it returns a coder that just doesn't work, that is a huge problem.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 12 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Yea, I'd be happy inlining this method, too.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 8 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: The duplication would come from mixing in a bunch of IOs. It should be rare, so not a huge deal. Just a best practice thing. I like the idea of re-registering bumping the priority.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java (hunk 6 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactories.java]: I see what you mean. If I call `registry.register(myClass, coderInstance)` I want to make sure that coder only handles `myClass`. So now that it is just a trial-and-move-on list we need some way to wrap it up like that. I would probably separate into `CoderFactoryForCoder(coderInstance)` and a separate `CoderFactoryOnlyForClass(myClass, delegateFactory)`. But all of this should be invisible to users anyhow so it isn't urgent. We should review the API surface after this is done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderFactoriesTest.java (hunk 17 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderFactoriesTest.java]: Done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/values/TypedPValueTest.java (hunk 22 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/values/TypedPValueTest.java]: As suggested, we can reduce the number of coder factories and list all their failure reasons. If users do enable many additional coder factories, then the error message will be very long.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java (hunk 19 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderRegistryTest.java]: It turns out the AvoCoder provider as it existed would say it can encode:
public static class X {
  public Thread t = Thread.currentThread();
}

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 15 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Sounds good. Just a list of coder factories is a starting point, and (your idea) putting detailed reasons in suppressed exceptions. And maybe later we can find a pipeline options hack to show those via command line.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Create.java (hunk 16 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Create.java]: Actually this is basically as good as status quo and we can make any changes backwards compatibly. And the feature is not really type safe to begin with. So scrap this comment.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/SerializableCoderTest.java [hunk 25]
```diff
@@ -230,4 +231,14 @@ public void testEncodedTypeDescriptor() throws Exception {
         SerializableCoder.of(MyRecord.class).getEncodedTypeDescriptor(),
         Matchers.equalTo(TypeDescriptor.of(MyRecord.class)));
   }
+
+  private static class AutoRegistration implements Serializable {
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/SerializableCoderTest.java]: More like "NonAutoRegistration", am I right?

**Comment by kennknowles** [comment]
Feel free to curate your commits and force push. I've got enough context, and GitHub has improved in this regard.

**Comment by lukecwik** [comment]
I dropped all the big renames and most of the removal of getInstanceComponents.

Please review again and once we are happy with the contents of the code, I can apply a bunch of renames on top to get us to where this original PR + your additional comments take us.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java (hunk 2 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java]: This rename matters.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactory.java [hunk 26]
```diff
@@ -18,27 +18,23 @@
...
 
 /**
- * A {@link CoderFactory} creates coders and decomposes values.
- * It may operate on a parameterized type, such as {@link List},
- * in which case the {@link #create} method accepts a list of
- * coders to use for the type parameters.
+ * A {@link CoderFactory} creates coders.
+ *
+ * <p>It may operate on a parameterized type, such as {@link List}, in which case the
+ * {@link #create} method accepts a list of coders to use for the type parameters.
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderFactory.java]: I guess if users are going to implement `CoderFactory` we should convert to abstract class or provide a base class that we tell them to extend. It makes me sad, since interfaces are so much more flexible in other ways and are really the right tool, but... Java.

**Comment by kennknowles** [comment]
I am happy with the implementation; let's get the API polished and the commit history clean.

**Comment by kennknowles** [comment]
Took a quick stab and fixing this, but it does turn out to have required foresight. Go ahead and squash and merge. LGTM.
